README.md
[![R-CMD-check](https://github.com/kolesarm/ebci/workflows/R-CMD-check/badge.svg)](https://github.com/kolesarm/ebci/actions) [![Coverage status](https://codecov.io/gh/kolesarm/ebci/branch/master/graph/badge.svg)](https://codecov.io/github/kolesarm/ebci?branch=master)

# ebci

This R package implements robust empirical Bayes confidence intervals from
[Armstrong, Kolesár, and Plagborg-Møller
(2020)](https://arxiv.org/abs/2004.03448). See the package
[manual](doc/manual.pdf) for documentation of the package functions, and the
[package vignette](doc/ebci.pdf) for a description of the package and an example
of its usage (available through `vignette("ebci")` once the package is
installed). See [ebci_matlab](https://github.com/mikkelpm/ebci_matlab) for a
Matlab version of this package, and
[ebciStata](https://github.com/kolesarm/ebciStata) for a Stata version.

This software package is based upon work supported by the National Science
Foundation under grant numbers SES-2049765 (Armstrong), SES-22049356 (Kolesár),
and (SES-1851665), and by work supported by the Alfred P. Sloan Research
Fellowship (Kolesár).

## Installation

You can get the current development version from GitHub:

``` r
if (!requireNamespace("remotes")) {
  install.packages("remotes")
}
remotes::install_github("kolesarm/ebci")
```

## Example

Calculation of the critical values used to construct robust empirical Bayes
confidence intervals:

``` r
library("ebci")
## Usual critical value
cva(m2=0, kappa=Inf, alpha=0.05)
## Larger critical value that takes bias into account. Only uses second moment
## constraint on normalized bias, m2=4.
cva(m2=4, kappa=Inf, alpha=0.05)
## Add a constraint that kurtosis equals 3. This tightens the critical value
cva(m2=4, kappa=3, alpha=0.05)
```

Estimates and robust EBCIs for neighborhood effects, as in the empirical
application in [Armstrong, Kolesár, and Plagborg-Møller
(2020)](https://arxiv.org/abs/2004.03448). Shrink fixed-effect estimates of the
neighborhood effects, for children with parents at the 25th percentile of the
income distribution (`theta25`) toward average outcome for permanent residents
(stayers) at the 25th percentile of the income distribution. Use precision
weights proportional to the inverse of the squared standard error of the
fixed-effect estimates.

``` r
r <- ebci(theta25 ~ stayer25, data=cz, se=se25, weights=1/se25^2)
```


NEWS.md
# ebci 1.0.0

## New Features

- The function `cva` computes the critical value derived in [Armstrong, Kolesár,
  and Plagborg-Møller (2020)](https://arxiv.org/abs/2004.03448) for constructing
  robust empirical Bayes confidence intervals.
- The function `ebci` computes empirical Bayes shrinkage estimators and
  associated parametric or robust empirical Bayes confidence intervals.


tests/testthat.R
library(testthat)
library(ebci)

test_check("ebci")


tests/testthat/test-cv.R
context("Check critical value calculations")

test_that("Numerical issues with rt0", {

    ## Convert return value from list to vector
    rt00 <- function(chi) unlist(rt0(chi))

    chis <- sqrt(3)+ exp(seq(log(1e-12), log(1e-3), length.out=200))
    f0 <- vapply(chis, rt00, numeric(2))
    expect_equal(sum(f0[1, ]<f0[2, ]), 0L)

    chis <- 10^seq(0, 5, length.out=200)
    f1 <- vapply(chis, rt00, numeric(2))
    expect_equal(sum(f1[1, ]<f1[2, ]), 0L)

    ## For large values of m2, Chebyshev inequality should be sharp
    m2 <- 1/1.11022e-16
    expect_equal(cva(m2)$cv/sqrt((1+m2)/0.05), 1L)
    expect_equal(cva(m2/100)$cv/sqrt((1+m2/100)/0.05), 1L)

    ## Assume kappa constraint not binding for large m2
    expect_warning(cv0 <- cva(m2, kappa=3, check=FALSE)$cv)
    expect_warning(rho(m2, kappa=3, chi=cv0))

    m2s <- seq(1/1000, 100, length.out=10)*m2
    vals <- vapply(m2s,
           function(r) cva(r, kappa=400, check=FALSE, alpha=0.1)$cv,
           numeric(1))
    expect_equal(vals/sqrt((1+m2s)/0.1), rep(1, length(m2s)))

    ## This is increasing, maximum at 4.8518e-26, close to 1
    expect_lt(abs(lam(x0=180144474255572736,
                     chi=424434295.36931574345)$lam/4.851805e-26 - 1L), 1e-6)
    ## This one is decreasing, max should be at 0
    expect_equal(lam(x0=12577.803781, chi=109.857326)$x0, 0L)
    ## First derivative evaluates to negative, but should probably be possitive;
    ## this shouldn't throw us off
    expect_equal(lam(x0 = 0.000759676644983789, chi = 2.85710241617874)$lam,
                 0.00827903)


})

test_that("Simple critical value sanity checks", {
    expect_equal(cva(m2=0, kappa=1)$cv, stats::qnorm(0.975))
    expect_equal(cva(m2=0, kappa=4)$cv, stats::qnorm(0.975))
    expect_equal(cva(m2=0, kappa=4, alpha=0.2)$cv, stats::qnorm(0.9))
    expect_equal(cva(m2=0, kappa=Inf)$cv, stats::qnorm(0.975))
    expect_equal(cva(m2=1, kappa=1)$cv, CVb(B=1))
    expect_equal(cva(m2=100^2, kappa=1)$cv, CVb(B=100))
    expect_equal(cva(m2=0, kappa=1)$cv, CVb(B=0))
    expect_equal(rho(m2=4, kappa=1,
                     chi=cva(m2=4, kappa=1)$cv)$alpha, 0.05)

    expect_equal(cva(m2=1, kappa=10000, check=TRUE, alpha=0.2)$cv,
                 cva(m2=1, kappa=Inf, check=TRUE, alpha=0.2)$cv)
    ## Here under kappa=Inf, solution has kurtosis (sol$x[2]^2*sol$p[2])/16=17.4
    expect_equal(cva(m2=4, kappa=20, check=TRUE, alpha=0.05)$cv,
                 cva(m2=4, kappa=Inf, check=TRUE, alpha=0.05)$cv)
    ## Here under kappa=Inf, solution has kurtosis (sol$x[2]^2*sol$p[2])/16=16.6
    expect_lt(cva(m2=0.25, kappa=15, check=TRUE, alpha=0.05)$cv,
                 cva(m2=0.25, kappa=Inf, check=TRUE, alpha=0.05)$cv)

    expect_equal(cva(m2=25, kappa=3, check=TRUE)$cv, 11.88358367)
    expect_equal(cva(m2=1, kappa=3, check=TRUE, alpha=0.2)$cv, 1.8494683)
    expect_lt(abs(cva(m2=25, kappa=1.000001)$cv-CVb(B=5)), 1e-5)
    expect_lt(abs(cva(m2=0.01^2, kappa=15)$cv-CVb(B=0.01)), 1e-5)
    expect_lt(abs(cva(m2=0.01^2,
                      kappa=15, alpha=0.2)$cv-CVb(B=0.01, alpha=0.2)),
              1e-5)

    ## Test LF distribution
    r1 <- cva(m2=4, kappa=1.000001, alpha=0.1)
    r2 <- cva(m2=4, kappa=1L, alpha=0.1)
    expect_lt(r1$x[which.max(r1$p)]-r2$x[which.max(r2$p)], 1e-3)
    expect_lt(abs(r1$p[which.max(r1$p)]-1), 1e-5)
})

test_that("Check large values", {
    expect_equal(lam(0, 42.2201)$x0, 1932.78377442)
    expect_equal(cva(10^2, 3)$cv, 24.86172217)
    expect_equal(rho(m2=1, kappa=3, chi=11.9699639845401)$alpha, 6.0216e-05)
    ## Previously this was failing
    expect_equal(cva(m2=2.8^2, alpha=0.1, kappa=8)$cv, 7.0779747158)
    expect_equal(rho(2.8^2, 8, 7.1324290147839279896)$alpha, 0.09805099)

    store_cva <- function(m2, kappa, alpha, check=TRUE) {
        df <- expand.grid(m2=m2, kappa=kappa, alpha=alpha)
        cvj <- function(j)
            cva(df$m2[j], kappa=df$kappa[j], df$alpha[j], check)$cv
        df$cv <- vapply(seq_len(nrow(df)), cvj, numeric(1))
        df
    }
    m2 <-  seq(0, 5, by=1)^2
    ## Takes 30s on X1 Carbon
    kappa <- c(1:10, Inf)
    expect_silent(store_cva(m2, kappa, alpha=0.1))
})

test_that("Check delta1 has crosses zero at most once", {
    checkdelta <- function(chi) {
        ts <- sort(unlist(rt0(chi)))
        xs <-  c(seq(0, ts[1], length.out=100),
                 seq(ts[1], ts[2], length.out=100))
        x0s <- xs
        ## If chi <=2.856, then derivative should always be negative
        if (chi <=  2.856) {
            expect_equal(sum(vapply(x0s, function(x0)
                sum(delta1(xs, x0, chi) >= 0), numeric(1))),
                0L)
        } else {
            ## Derivative should be first positive, then negative. Only need
            ## this to hold on [0, t1] if x>t1
            der <- function(x0) {
                    delta1(xs[1:(100*(x0 >= ts[1])+(x0<ts[1])*200)], x0, chi)
            }
            expect_equal(sum(vapply(x0s, function(x0)
                is.unsorted(!(der(x0) >= 0)), logical(1))),
                0L)
        }
    }
    chis <- seq(sqrt(3), 50, length.out=300)
    vapply(chis, checkdelta, numeric(1))
})


tests/testthat/test-eb.R
context("Check optimal and EB shrinkage")

test_that("Table 1 in paper", {
    ts <- cz[!is.na(cz$theta25), ]
    eb_table <- function(res, wgt) {
        res$df$sn <- res$mu2[1]/res$df$se^2
        l1 <- vapply(res$df, weighted.mean, FUN.VALUE=numeric(1), w=wgt)
        ret <- c("sqrt{mu_2}"=sqrt(res$mu2[1]),
                 "kappa"=res$kappa[1],
                 l1[length(l1)],
                 res$delta, l1[1:7], l1[4]/l1[5:7])
        ret
    }
    ## Only large CZ to  make it faster
    df <- cz[sort(cz$pop, index.return=TRUE, decreasing=TRUE)$ix[1:10], ]
    i2 <- ebci(formula=theta25~stayer25, data=df, se=se25, alpha=0.1, wopt=TRUE,
               fs_correction="none")
    i3 <- ebci(formula=theta25~0, data=df, se=se25, alpha=0.05, kappa=Inf,
               wopt=TRUE)

    testthat::expect_equal(unname(round(eb_table(i2), 6)),
                           c(0.060852, 5.959142, 0.956639, -0.913932, 0.020049,
                             0.456490, 0.526469, 0.115947, 0.078879, 0.075404,
                             0.073284, 0.113034, 1.046083, 1.076356, 0.697836))
    testthat::expect_equal(unname(round(eb_table(i3), 6)),
                           c(0.085168, Inf, 1.873921, 0.612815, 0.695995,
                             0.063068, 0.114249, 0.107078, 0.102651, 0.134688,
                             1.066973, 1.112989, 0.848247))
})


test_that("ebci function problems", {
    ## negative mu2 estimate
    df <- cz[sort(cz$pop, index.return=TRUE, decreasing=TRUE)$ix[1:100], ]
    expect_warning(r <- ebci(formula=theta25~stayer25, data=df, kappa=Inf,
                             se=se25, alpha=0.1, fs_correction="none"))
})

test_that("Zero bias estimate", {
    expect_identical(cva(m2=Inf, kappa=1)$cv, NA)
    expect_warning(r <- w_opt(S=0, kappa=3))
    expect_true(is.na(r$length))
    expect_true(is.na(w_eb(S=0, kappa=1)$length))

})

test_that("fs corrections", {
    df <- cz[sort(cz$pop, index.return=TRUE, decreasing=TRUE)$ix[1:40], ]
    r <- ebci(formula=theta25~stayer25, data=df, se=se25,
              alpha=0.1, fs_correction="FPLIB", wopt=TRUE)
    r2 <- ebci(formula=theta25~stayer25, data=df, se=se25,
              alpha=0.1, wopt=FALSE, fs_correction="PMT")
    expect_equal(unname(c(sqrt(r$mu2[1]), r$kappa[1])),
                 c(0.04386307, 34.60536787))
    expect_equal(as.vector(summary(r$df$len_op)),
                 c(0.0510135245, 0.06513867, 0.071050167, 0.069886713,
                   0.076510241, 0.080472608))
    expect_equal(unname(sqrt(r2$mu2)), c(0.0247533358, 0.016287771))
    expect_equal(unname(r2$kappa), c(475.964641182, 1719.44225761))

    ## When noise is small, FPLIB should be the similar
    set.seed(42)
    th <- 42+rnorm(500)/4
    se <- sqrt(seq(1, 10, length.out=500))/10
    Y <- th + rnorm(500)*se
    rt <- ebci(formula=Y~1, se=se, weights=1/se^2, wopt=FALSE)
    rn <- ebci(formula=Y~1, se=se, weights=1/se^2, wopt=FALSE,
               fs_correction = "none")
    rb <- ebci(formula=Y~1, se=se, weights=1/se^2, wopt=FALSE,
               fs_correction = "FPLIB")
    expect_identical(c(rt$kappa, rt$mu2), c(rn$kappa, rn$mu2))
    expect_lt(max(abs(c(rb$kappa, rb$mu2)-c(rn$kappa, rn$mu2))), 1e-2)
    est <- colSums(abs(rt$df-rb$df))
    expect_lt(max(est[!is.na(est)]), 1e-2)
    ## Test format output while we're at it
    expect_identical(dim(rt$df), c(500L, 13L))
    expect_identical(dim(rb$df), c(500L, 13L))
    rtt <- ebci(formula=Y~1, se=se, weights=1/se^2, wopt=TRUE,
               fs_correction = "FPLIB")
    expect_identical(dim(rt$df), dim(rtt$df))
})

test_that("Pre-computing critical values", {
    dfm <- data.frame(
        y=c(-0.0912394115425668, 0.270659035981612, 0.0505183828184417,
            0.401285011472904, -0.116494855854479, -0.334314741812537,
            -0.18928733418011, 0.165181263555627, 0.132184880945018,
            -0.131267800385369, 0.584675521401188, 0.105172038899963,
            0.470647126771099, 0.289695479626726, 0.202471324044271,
            0.0146653486070987, 0.568678856504733, 0.216916114930276,
            -0.259804655342694, -0.582460577829165, -0.449075030332067,
            -0.251549682686536, -0.331692925639294, -0.116896950204574,
            -0.175038163494127, 0.178108950152456, 0.211683997890061,
            -0.911066040123521, -0.164931392459329, -0.161697036212087,
            -0.0662609931760586, 0.0316248074589969),
        se=c(0.224386445769098, 0.257751786695356, 0.276711829264707,
             0.403355134534814, 0.343841207609935, 0.260088725292071,
             0.239258750261168, 0.177225564240537, 0.165070998037012,
             0.215345806604199, 0.329392513934246, 0.166990070498265,
             0.356719957739564, 0.338361696727126, 0.26596125123086,
             0.278943346050078, 0.286066216054302, 0.211277847211765,
             0.197890061668623, 0.254834949524809, 0.161320618248102,
             0.159717394888581, 0.190422698118735, 0.11198110341991,
             0.128750097633114, 0.137084993056956, 0.144134717528643,
             0.198405131349222, 0.235877391582392, 0.129194532604923,
             0.0592319350677268, 0.45161460229795))
    ## Should be no warning, previously we did get them
    expect_silent(c2 <- ebci(y ~ 0, dfm, se, wopt=TRUE, alpha=0.05, kappa=Inf))
    expect_equal(dim(c2$X), c(32L, 0L))
})


R/document_data.R
#' Neighborhood effects data from Chetty and Hendren (2018)
#'
#' This dataset contains a subset of the publicly available data from Chetty and
#' Hendren (2018). It contains raw estimates and standard errors of neighborhood
#' effects at the commuting zone level
#' @format A data frame with 741 rows corresponding to commuting zones (CZ) and
#'     10 columns corresponding to the variables:
#'
#' \describe{
#'
#' \item{cz}{Commuting zone ID}
#'
#' \item{czname}{Name of CZ}
#'
#' \item{state}{2-digit state code}
#'
#' \item{pop}{Population according to the year 2000 Census}
#'
#' \item{theta25}{Fixed-effect estimate of the causal effect of living in the CZ
#' for one year on children's percentile rank in the national distribution of
#' household earnings at age 26 relative to others in the same birth cohort for
#' children growing up with parents at the 25th percentile of national income
#' distribution}
#'
#' \item{theta75}{Fixed-effect estimate of the causal effect of living in the CZ
#' for one year on children's percentile rank in the national distribution of
#' household earnings at age 26 relative to others in the same birth cohort for
#' children growing up with parents at the 75th percentile of national income
#' distribution}
#'
#' \item{se25}{Standard error of \code{theta25}}
#'
#' \item{se75}{Standard error of \code{theta75}}
#'
#' \item{stayer25}{Average percentile rank in the national distribution of
#' household earnings at age 26 relative to others in the same birth cohort for
#' stayers (children who grew up in the CZ and did not move) with parents at the
#' 25th percentile of national income distribution.}
#'
#' \item{stayer75}{Average percentile rank in the national distribution of
#' household earnings at age 26 relative to others in the same birth cohort for
#' stayers (children who grew up in the CZ and did not move) with parents at the
#' 75th percentile of national income distribution.}
#'
#' }
#' @source \url{https://opportunityinsights.org/data/?paper_id=599}

#' @references{
#'
#' \cite{Chetty, R., & Hendren, N. (2018). The Impacts of Neighborhoods on
#' Intergenerational Mobility II: County-Level Estimates. The Quarterly Journal
#' of Economics, 133(3), 1163–1228. \doi{10.1093/qje/qjy006}}
#'
#' }
"cz"


R/eb.R
## #' Optimal shrinkage for Empirical Bayes confidence intervals
## #'
## #' Compute linear shrinkage factor \eqn{w_{opt}} to minimize the length
## #' of the resulting Empirical Bayes confidence interval (EBCI).
## #' @param S Signal-to-noise ratio \eqn{\mu_{2}/\sigma^2}{mu_2/sigma^2}, where
## #'     \eqn{\mu_{2}}{mu_2} is the second moment of
## #'     \eqn{\theta-X'\delta}{theta-X*delta}, and \sigma^2 is the
## #'     variance of the preliminary estimator.
## #' @param kappa Kurtosis of \eqn{\theta-X'\delta}{theta-X*delta}.
## #' @param alpha Determines confidence level, \eqn{1-\alpha}{1-alpha}.
## #' @param cv_tbl Optionally, supply a data frame of critical values.
## #'     \code{cva(m2, kappa, alpha)}, for different values of \code{m2}, such
## #'     as a subset of the data frame \code{\link{cva_tbl}}, that matches the
## #'     supplied values of \code{alpha} and \code{kappa}. The data frame needs
## #'     to contain two variables, \code{m2}, corresponding to the value of 2nd
## #'     moment of the normalized bias, and \code{cv}, with the corresponding
## #'     value of \code{cva(m2, kappa, alpha)}. If non \code{NULL}, for the
## #'     purposes of optimizing the shrinkage factor, compute the
## #'     \code{cva} by interpolating between the critical values in this data
## #'     frame, instead of computing them from scratch. This can speed up the
## #'     calculations.
## #' @return Returns a list with 3 components: \describe{
## #'
## #' \item{\code{w}}{Optimal shrinkage factor \eqn{w_{opt}}{w_opt}}
## #'
## #' \item{\code{length}}{Normalized half-length of the corresponding conf
## #' interval, so that the interval obtains by taking the estimator based on
## #' shrinkage given by \code{w}, and adding and subtracting \code{length}
## #' times the standard error \eqn{\sigma}{sigma} of preliminary estimator.}
## #'
## #' \item{\code{m2}}{The second moment of the normalized bias,
## #' \eqn{(1/w-1)^{2}S}{(1/w-1)^2*S}}}
## #' @references{
## #'
## #' \cite{Armstrong, Timothy B., Kolesár, Michal, and Plagborg-Møller, Mikkel
## #' (2020): Robust Empirical Bayes Confidence Intervals,
## #' \url{https://arxiv.org/abs/2004.03448}}
## #'
## #' }
## #' @examples
## #' w_opt(4, 3)
## #' ## Use precomputed critical value table
## #' w_opt(4, 3, cv_tbl=cva_tbl[cva_tbl$kappa==3 & cva_tbl$alpha==0.05, ])
w_opt <- function(S, kappa, alpha=0.05, cv_tbl=NULL) {
    if (!is.null(cv_tbl)) {
        ## Linearly interpolate
        cv <- function(m2) {
            idx <- which.max(cv_tbl$m2>m2)
            idx0 <- max(idx-1, 1)
            cv_tbl$cv[idx0]+(m2-cv_tbl$m2[idx0])/
                (cv_tbl$m2[idx]-cv_tbl$m2[idx0]) *
                (cv_tbl$cv[idx]-cv_tbl$cv[idx0])
        }
        maxbias <- max(cv_tbl$m2)
        minbias <- min(cv_tbl$m2)
    } else {
        cv <- function(m2) cva(m2, kappa=kappa, alpha=alpha, check=FALSE)$cv
        ## Maxium bias (i.e m2) is 100^2 to prevent numerical issues
        maxbias <- 100^2
        minbias <- 0
    }
    ci_length <- function(w) cv((1/w-1)^2*S)*w
    if (S > tol) {
        r <- stats::optimize(ci_length, c(1/(sqrt(maxbias/S)+1),
                                          1/(sqrt(minbias/S)+1)), tol=tol)
        m2 <- (1/r$minimum-1)^2*S
    } else {
        r <- list(minimum=0, objective=NA)
        m2 <- Inf
    }
    ## Recompute critical value, checking solution accuracy. If we're using
    ## cv_tbl, then optimum will be at one of the values of m2 in the table, so
    ## solution should always be accurate
    len <- cva(m2, kappa=kappa, alpha=alpha, check=TRUE)$cv*r$minimum
    if (m2 > maxbias-1e-4)
        warning("Corner solution: optimum reached at maximal normalized bias")
    list(w=r$minimum, length=len, m2=m2)
}

## #' Empirical Bayes estimator and confidence intervals
## #'
## #' Compute empirical Bayes shrinkage factor \eqn{w_{eb}}{w_eb} and the
## #' normalized half-length of of the associated robust Empirical Bayes
## #' confidence interval (EBCI).
## #' @inheritParams w_opt
## #' @return Returns a list with 3 components:
## #' \describe{
## #'
## #' \item{\code{w}}{Empirical Bayes shrinkage factor \eqn{w_{eb}}{w_eb}}
## #'
## #' \item{\code{length}}{Normalized half-length of the corresponding conf
## #' interval, so that the interval obtains by taking the estimator based on
## #' shrinkage given by \code{w}, and adding and subtracting \code{length}
## #' times the standard error \eqn{\sigma}{sigma} of preliminary estimator.}
## #'
## #' \item{\code{m2}}{Second moment of the normalized bias, \code{1/m2}.}
## #'
## #' }
## #' @references{
## #'
## #' \cite{Armstrong, Timothy B., Kolesár, Michal, and Plagborg-Møller, Mikkel
## #' (2020): Robust Empirical Bayes Confidence Intervals,
## #' \url{https://arxiv.org/abs/2004.03448}}
## #'
## #' }
## #' @examples
## #' w_eb(2, 3)
## #' ## No constraint on kurtosis yields doesn't affect shrinkage, but yields
## #' ## larger half-length
## #' w_eb(2, Inf)
w_eb <- function(S, kappa=Inf, alpha=0.05) {
    w <- S/(1+S)
    list(w=w, length=cva(1/S, kappa=kappa, alpha=alpha, check=TRUE)$cv*w,
         m2=1/S)
}

## (Trimmed) moments estimation
moments <- function(eps, se, wgt, fs_correction="PMT", mu2=NULL, kappa=NULL) {
    wgtV <- function(Z, wgt)
        sum(wgt^2*(Z^2-stats::weighted.mean(Z, wgt)^2))/(sum(wgt)^2-sum(wgt^2))
    tmean <- function(m, V)  m + sqrt(V)*stats::dnorm(m/sqrt(V))/
                                 stats::pnorm(m/sqrt(V))

    w2 <- eps^2 - se^2
    w4 <- eps^4 - 6*se^2*eps^2 + 3*se^4
    ## \tilde{mu}_2, \tilde{mu}_4)
    tmu <- c(stats::weighted.mean(w2, wgt), stats::weighted.mean(w4, wgt))
    pmt_trim <- c(2 *mean(wgt^2*se^4) / (sum(wgt)*mean(wgt*se^2)),
                  32*mean(wgt^2*se^8) / (sum(wgt)*mean(wgt*se^4)))
    if (is.null(mu2))
        mu2 <- switch(fs_correction,
                      "none"=max(tmu[1], 0),
                      "PMT"=max(tmu[1], pmt_trim[1]),
                      "FPLIB"=tmean(tmu[1], wgtV(w2, wgt)),
                      stop("Unknown fs_correction method"))
    if (is.null(kappa))
        kappa <- switch(fs_correction,
                        "none"=max(tmu[2]/mu2^2, 1),
                        "PMT"=max(tmu[2]/mu2^2, 1 + pmt_trim[2]/mu2^2),
                        "FPLIB"= 1+tmean(tmu[2]-tmu[1]^2,
                                         wgtV(w4-2*mu2*w2, wgt))/mu2^2)
    c(mu2, kappa, tmu[1], tmu[2]/tmu[1]^2)
}

#' Compute empirical Bayes confidence intervals by shrinking toward regression
#'
#' Computes empirical Bayes estimators based on shrinking towards a regression,
#' and associated robust empirical Bayes confidence intervals (EBCIs), as well
#' as length-optimal robust EBCIs.
#' @param formula object of class \code{"formula"} (or one that can be coerced
#'     to that class) of the form \code{Y ~ predictors}, where \code{Y} is a
#'     preliminary unbiased estimator, and \code{predictors} are predictors
#'     \eqn{X} that guide the direction of shrinkage. For shrinking toward the
#'     grand mean, use \code{Y ~ 1}, and for shrinking toward \code{0} use
#'     \code{Y ~ 0}
#' @param data optional data frame, list or environment (or object coercible by
#'     \code{as.data.frame} to a data frame) containing the preliminary
#'     estimator \code{Y} and the predictors. If not found in \code{data}, these
#'     variables are taken from \code{environment(formula)}, typically the
#'     environment from which the function is called.
#' @param se Standard errors \eqn{\sigma}{sigma} associated with the preliminary
#'     estimates \code{Y}
#' @param weights An optional vector of weights to be used in the fitting
#'     process in computing \eqn{\delta}{delta}, \eqn{\mu_2}{mu_2} and
#'     \eqn{\kappa}{kappa}. Should be \code{NULL} or a numeric vector.
#' @param alpha Determines confidence level, \eqn{1-\alpha}{1-alpha}.
#' @param kappa If non-\code{NULL}, use pre-specified value for the kurtosis
#'     \eqn{\kappa}{kappe} of \eqn{\theta-X'\delta}{theta-X*delta} (such as
#'     \code{Inf}), instead of computing it.
#' @param fs_correction Finite-sample correction method used to compute
#'     \eqn{\mu_2}{mu_2} and \eqn{\kappa}{kappa}. These corrections ensure that
#'     we do not shrink the preliminary estimates \code{Y} all the way to zero.
#'     If \code{"PMT"}, use posterior mean truncation, if \code{"FPLIB"} use
#'     limited information Bayesian approach with a flat prior, and if
#'     \code{"none"}, truncate the estimates at \code{0} for \eqn{\mu_2}{mu_2}
#'     and \code{1} for \eqn{\kappa}{kappa}.
#' @param wopt If \code{TRUE}, also compute length-optimal robust EBCIs. These
#'     are robust EBCIs centered at estimates with the shrinkage factor
#'     \eqn{w_i}{w_i} chosen to minimize the length of the resulting EBCI.
#' @return Returns a list with the following components: \describe{
#'
#' \item{\code{mu2}}{Estimated second moment of
#' \eqn{\theta-X'\delta}{theta-X*delta}, \eqn{\mu_2}{mu_2}. Vector
#' of length 2, the first element corresponds to the estimate after the
#' finite-sample correction as specified by \code{fs_correction}, the second
#' element is the uncorrected estimate.}
#'
#' \item{\code{kappa}}{Estimated kurtosis \eqn{\kappa}{kappa} of
#' \eqn{\theta-X'\delta}{theta-X*delta}. Vector of length 2 with the same
#' structure as \code{mu2}.}
#'
#' \item{\code{delta}}{Estimated regression coefficients \eqn{\delta}{delta}}
#'
#' \item{\code{X}}{Matrix of regressors}
#'
#' \item{\code{alpha}}{Determines confidence level \eqn{1-\alpha} used.}
#'
#' \item{\code{df}}{Data frame with components described below.}
#' }
#'
#' \code{df} has the following components:
#'
#' \describe{
#'
#' \item{\code{w_eb}}{EB shrinkage factors,
#'    \eqn{\mu_{2}/(\mu_{2}+\sigma^2_i)}{mu_2/(mu_2+sigma^2_i)}}
#'
#' \item{\code{w_opt}}{Length-optimal shrinkage factors}
#'
#' \item{\code{ncov_pa}}{Maximal non-coverage of parametric EBCIs}
#'
#' \item{\code{len_eb}}{Half-length of robust EBCIs based on EB shrinkage, so
#' that the intervals take the form \code{cbind(th_eb-len_eb, th_eb+len_eb)}}
#'
#' \item{\code{len_op}}{Half-length of robust EBCIs based on length-optimal
#' shrinkage, so that the intervals take the form \code{cbind(th_op-len_op,
#' th_op+len_op)}}
#'
#' \item{\code{len_pa}}{Half-length of parametric EBCIs, which take the form
#' \code{cbind(th_eb-len_pa, th_eb+len_a)}}
#'
#' \item{\code{len_us}}{Half-length of unshrunk CIs, which take the form
#' \code{cbind(th_us-len_us, th_us+len_us)}}
#'
#' \item{\code{th_us}}{Unshrunk estimate \eqn{Y}}
#'
#' \item{\code{th_eb}}{EB estimate.}
#'
#' \item{\code{th_op}}{Estimate based on length-optimal shrinkage.}
#'
#' \item{\code{se}}{Standard error \eqn{\sigma}{sigma}, as supplied by the
#' argument \code{se}}
#'
#' \item{\code{weights}}{Weights used}
#'
#' \item{\code{residuals}}{The residuals \eqn{Y_i-X_i\delta}{Y_i-X_i*delta}}
#'
#' }
#' @references{
#'
#' \cite{Armstrong, Timothy B., Kolesár, Michal, and Plagborg-Møller, Mikkel
#' (2020): Robust Empirical Bayes Confidence Intervals,
#' \url{https://arxiv.org/abs/2004.03448}}
#'
#' }
#' @examples
#' ## Same specification as in empirical example in Armstrong, Kolesár
#' ## and Plagborg-Møller (2020), but only use data on NY commuting zones
#' r <- ebci(theta25 ~ stayer25, data=cz[cz$state=="NY", ],
#'           se=se25, weights=1/se25^2)
#' @export
ebci <- function(formula, data, se, weights=NULL, alpha=0.1, kappa=NULL,
                 wopt=FALSE, fs_correction="PMT") {
    ## Construct model frame
    mf <- match.call(expand.dots = FALSE)
    m <- match(c("formula", "data", "se", "weights"), names(mf), 0L)
    mf <- mf[c(1L, m)]
    mf$formula <- stats::as.formula(formula)
    mf[[1L]] <- quote(stats::model.frame)
    mf <- eval(mf, parent.frame())
    Y <- stats::model.response(mf, "numeric")
    wgt <- as.vector(stats::model.weights(mf))
    ## Do not weight
    if (is.null(wgt))
        wgt <- rep(1L, length(Y))
    se <- mf$"(se)"
    X <- stats::model.matrix(attr(mf, "terms"), mf)
    za <- stats::qnorm(1-alpha/2)

    ## Compute delta, and moments
    r1 <- stats::lm.wfit(X, Y, wgt)
    mu1 <- unname(r1$fitted.values)
    delta <- r1$coefficients

    ## Weighted sample variance and mean of truncated normal
    mus <- moments(Y-mu1, se, wgt, fs_correction, kappa=kappa)

    if (mus[1] == 0) {
        warning("mu2 estimate is 0")
        df <- data.frame(w_eb=se*0, w_opt=se*0, ncov_pa=se*NA, len_eb=se*NA,
                         len_op=se*NA, len_pa=se*NA, len_us=za*se, th_us=Y,
                         th_eb=Y-mu1, th_op=Y-mu1, se=se)
        return(list(mu2=c("estimate"=mus[1], "uncorrected_estimate"=mus[3]),
                    kappa=c("estimate"=mus[2], "uncorrected_estimate"=mus[4]),
                    delta=delta, df=df, X=X, alpha=alpha))
    }

    ## EB shrinkage
    eb <- vapply(se, function(se) unlist(w_eb(mus[1]/se^2, mus[2], alpha)),
                 numeric(3))
    eb <- as.data.frame(t(eb))
    th_eb <- mu1+eb$w*(Y-mu1)

    ## Optimal shrinkage. First pre-compute cva for this kurtosis
    if (length(se) > 30 & wopt) {
        mmin <- w_opt(mus[1]/min(se)^2, mus[2], alpha=alpha)$m2
        mmax <- w_opt(mus[1]/max(se)^2, mus[2], alpha=alpha)$m2
        df <- data.frame(m2=c(mmin*0.95, seq(mmin, mmax, length.out=501),
                              mmax*1.05))
        cvj <- function(j)
            cva(df$m2[j], kappa=mus[2], alpha, check=TRUE)$cv
        df$cv <- vapply(seq_len(nrow(df)), cvj, numeric(1))
    } else {
        df <- NULL
    }

    if (wopt) {
        op <- vapply(se, function(se)
            unlist(w_opt(mus[1]/se^2, mus[2], cv_tbl=df,
                         alpha=alpha)), numeric(3))
        op <- as.data.frame(t(op))
        th_op <- mu1+op$w*(Y-mu1)
    } else {
        op <- eb*NA
        th_op <- th_eb*NA
    }

    ## Nonparametric moment estimates for critical values
    par_cov <- function(j)
        rho(se[j]^2/mus[1], mus[2],
            za*sqrt(1+se[j]^2/mus[1]), check=TRUE)$alpha
    ncov_pa <- vapply(seq_along(Y), par_cov, numeric(1))

    df <- data.frame(w_eb=eb$w, w_opt=op$w, ncov_pa=ncov_pa,
                     len_eb=eb$length*se, len_op=op$length*se,
                     len_pa=sqrt(eb$w)*za*se, len_us=za*se, th_us=Y,
                     th_eb=th_eb, th_op=th_op, se=se,
                     weights=wgt, residuals=Y-mu1)
    list(mu2=c("estimate"=mus[1], "uncorrected_estimate"=mus[3]),
         kappa=c("estimate"=mus[2], "uncorrected_estimate"=mus[4]),
         delta=delta, df=df, X=X, alpha=alpha)
}
